lahja/asyncio/endpoint.py

```diff
@@ -429,16 +434,25 @@ def subscribed_events(self) -> Set[Type[BaseEvent]]:
             .union(self._queues.keys())
         )
 
-    async def _notify_subscriptions_changed(self) -> None:
-        """
-        Tell all inbound connections of our new subscriptions
-        """
-        # make a copy so that the set doesn't change while we iterate over it
-        subscribed_events = self.subscribed_events
-        for remote in self._half_connections.copy():
-            await remote.notify_subscriptions_updated(subscribed_events)
-        for remote in tuple(self._full_connections.values()):
-            await remote.notify_subscriptions_updated(subscribed_events)
+    async def _monitor_subscription_changes(self) -> None:
+        self._subscriptions_changed = asyncio.Event()
+
+        while True:
+            # We wait for the event to change and then immediately replace it
+            # with a new event.  This **must** occur before any additional
+            # `await` calls to ensure that any *new* changes to the
+            # subscriptions end up operating on the *new* event and will be
+            # picked up in the next iteration of the loop.
+            await self._subscriptions_changed.wait()
+            self._subscriptions_changed = asyncio.Event()
+
+            # make a copy so that the set doesn't change while we iterate
+            # over it
+            subscribed_events = self.subscribed_events
```

Comment:
minor: I forgot this was a property and confused myself. Something like `subscribed_events = frozenset(self.subscribed_events)` would have saved me even though it's technically redundant. It's also slightly safer, the current code requires that `subscribed_events` is a property but does nothing to enforce that assumption.